File: 每日一题/797. 所有可能的路径.py
# https://leetcode-cn.com/problems/all-paths-from-source-to-target/
from typing import List


class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        path_ans = []
        n = len(graph)

        def dfs(nxt: int, p: List[int]):
            p.append(nxt)
            if nxt == n-1:
                path_ans.append(p[:])
                return
            for node in graph[nxt]:
                if node not in p:
                    dfs(node, p)
                    p.pop()

        dfs(0, [])
        return path_ans


x = Solution()
print(x.allPathsSourceTarget([[4, 3, 1], [3, 2, 4], [3], [4], []]))

File: labuladong/797. 所有可能的路径.js
/**https://leetcode-cn.com/problems/all-paths-from-source-to-target/
 * @param {number[][]} graph
 * @return {number[][]}
 */
var allPathsSourceTarget = function (graph) {
    const ret = [];
    function traverse(graph = [[]], s = 0, path = []) {
        path.push(s);
        const n = graph.length;
        // 到达终点
        if (s === n - 1) {
            ret.push(Array.from(path));
            path.pop();
            return;
        }
        // 访问相邻节点
        for (const v of graph[s]) {
            traverse(graph, v, path);
        }
        // 结束当前节点的访问
        path.pop();
    }
    traverse(graph, 0, []);
    return ret;
};
console.log(allPathsSourceTarget([[1, 2], [3], [3], []]));
